In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries imported.')

Libraries imported.


In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

# Part 1: Transform the data in the table into the pandas dataframe

In [3]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(website_url,'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className=document.documentElement.className.replace(/(^|\s)client-nojs(\s|$)/,"$1client-js$2");RLCONF={"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":906439794,"wgRevisionId":906439794,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":!1,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June",

In [4]:
My_table = soup.find('table',{'class':'wikitable sortable'})
My_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
</td></tr>
<tr>
<td>M6A</td>

In [5]:
columns = [th.text.replace('\n', '') for th in My_table.find('tr').find_all('th')]
columns  


['Postcode', 'Borough', 'Neighbourhood']

In [6]:
trs = My_table.find_all('tr')[1:]
rows = list()
for tr in trs:
    rows.append([td.text.replace('\n', '').replace('\xa0', '') for td in tr.find_all('td')])
rows[:5]

[['M1A', 'Not assigned', 'Not assigned'],
 ['M2A', 'Not assigned', 'Not assigned'],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Harbourfront']]

In [7]:
df_Toronto = pd.DataFrame(data=rows, columns=columns)

In [8]:
indexNames = df_Toronto[df_Toronto['Borough'] == 'Not assigned' ].index
print(indexNames)

Int64Index([  0,   1,   9,  13,  20,  21,  30,  36,  37,  45,  46,  50,  51,
             52,  54,  55,  59,  60,  61,  73,  74,  75,  88,  89,  90, 104,
            105, 106, 120, 121, 136, 137, 148, 149, 155, 161, 162, 167, 175,
            181, 182, 188, 189, 190, 194, 195, 201, 202, 203, 204, 209, 210,
            223, 224, 237, 238, 241, 242, 247, 248, 253, 254, 258, 259, 260,
            261, 263, 264, 274, 275, 276, 277, 278, 279, 280, 281, 287],
           dtype='int64')


In [9]:
df_Toronto.drop(indexNames, inplace=True)


In [10]:
df_Toronto = df_Toronto.groupby('Postcode').agg({'Borough':'first', 'Neighbourhood': ', '.join}).reset_index()

In [11]:
df_Toronto.loc[df_Toronto['Borough'] == "Queen's Park", ['Neighbourhood']] = "Queen's Park"


In [12]:
df_Toronto.shape

(103, 3)

# The end of Part 1: The no. of rows of my dataframe  is 103.

# Part 2 : Get the latitude & Longtitude of each neighborhood

In [13]:
LaLo = pd.read_csv('http://cocl.us/Geospatial_data')

In [14]:
df_Toronto.columns = ['Postal Code','Borough','Neighbourhood']

In [15]:
df_Toronto_LoLa = pd.merge(df_Toronto,LaLo, on = 'Postal Code')

In [16]:
df_Toronto_LoLa.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# The above is the result of Part 2: adding Latitude and Longitude

# Clustering Neighborhoods which are in boroughs that contain the word Toronto 

In [17]:
#select borough names containing Toronto
value_list = ['West Toronto', 'East Toronto', 'Downtown Toronto']
Toronto_data = df_Toronto_LoLa[df_Toronto_LoLa['Borough'].isin(value_list)].reset_index(drop=True)
Toronto_data.head(10)


Postal Code           Borough                   Neighbourhood   Latitude  \
0         M4E      East Toronto                     The Beaches  43.676357   
1         M4K      East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L      East Toronto  The Beaches West, India Bazaar  43.668999   
3         M4M      East Toronto                 Studio District  43.659526   
4         M4W  Downtown Toronto                        Rosedale  43.679563   
5         M4X  Downtown Toronto     Cabbagetown, St. James Town  43.667967   
6         M4Y  Downtown Toronto            Church and Wellesley  43.665860   
7         M5A  Downtown Toronto       Harbourfront, Regent Park  43.654260   
8         M5B  Downtown Toronto        Ryerson, Garden District  43.657162   
9         M5C  Downtown Toronto                  St. James Town  43.651494   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.377529  
5 -79.367675  
6 -79.383160  
7 -79.360636  
8 -79.378937  
9 -79.375418

In [28]:
#Foursquare
CLIENT_ID = '0PCSZ3NLJLPZKYIIEEHLZ4QTP0DSKAOGFGQYCWYKRQQ3QDM0' # your Foursquare ID
CLIENT_SECRET = 'QSUDJABPBL2RV2BRELXKCQRE334INB24TNNGMAA1BU2FPDIK' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 0PCSZ3NLJLPZKYIIEEHLZ4QTP0DSKAOGFGQYCWYKRQQ3QDM0
CLIENT_SECRET:QSUDJABPBL2RV2BRELXKCQRE334INB24TNNGMAA1BU2FPDIK


In [19]:
LIMIT = 50
radius = 500

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighbourhood'],latitudes=Toronto_data['Latitude'],longitudes=Toronto_data['Longitude'])

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The Junction South
Parkdale, Roncesvalles
Runnymede, Swansea
Business Reply Mail Processing Centre 969 Eastern


In [31]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1592, 7)


Neighbourhood  Neighbourhood Latitude  \
0                   The Beaches               43.676357   
1                   The Beaches               43.676357   
2                   The Beaches               43.676357   
3                   The Beaches               43.676357   
4  The Danforth West, Riverdale               43.679557   

   Neighbourhood Longitude                               Venue  \
0               -79.293031                   Glen Manor Ravine   
1               -79.293031  The Big Carrot Natural Food Market   
2               -79.293031                 Grover Pub and Grub   
3               -79.293031                       Upper Beaches   
4               -79.352188                            Pantheon   

   Venue Latitude  Venue Longitude     Venue Category  
0       43.676821       -79.293942              Trail  
1       43.678879       -79.297734  Health Food Store  
2       43.679181       -79.297215                Pub  
3       43.680563       -79.292869       Neighborhood  
4       43.677621       -79.351434   Greek Restaurant

In [32]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            23   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower, Bathurst Quay, Island airport, Harbou...                      16   
Cabbagetown, St. James Town                                             46   
Central Bay Street                                                      83   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    86   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
Dovercourt Village, Dufferin                                            20   
First Canadian Place, Underground city                                 100   
Harbord, University of Toronto                                          34   
Harbourfront East, Toronto Islands, Union Station                      100   
Harbourfront, Regent Park                                               50   
High Park, The Junction South                                           22   
Little Portugal, Trinity                                                63   
Parkdale, Roncesvalles                                                  15   
Rosedale                                                                 5   
Runnymede, Swansea                                                      39   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         94   
Studio District                                                         41   
The Beaches                                                              4   
The Beaches West, India Bazaar                                          20   
The Danforth West, Riverdale                                            42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide, King, Richmond                                                100   
Berczy Park                                                              55   
Brockton, Exhibition Place, Parkdale Village                             23   
Business Reply Mail Processing Centre 969 Eastern                        18   
CN Tower, Bathurst Quay, Island airport, Harbou...                       16   
Cabbagetown, St. James Town                                              46   
Central Bay Street                                                       83   
Chinatown, Grange Park, Kensington Market                               100   
Christie                                                                 16   
Church and Wellesley                                                     86   
Commerce Court, Victoria Hotel                                          100   
Design Exchange, Toronto Dominion Centre                                100   
Dovercourt Village, Dufferin                                             20   
First Canadian Place, Underground city                                  100   
Harbord, University of Toronto                                           34   
Harbourfront East, Toronto Islands, Union Station                       100   
Harbourfront, Regent Park                                                50   
High Park, The Junction South                                            22   
Little Portugal, Trinity                             

In [33]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 229 uniques categories.


In [34]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  \
0                   The Beaches                  0        0   
1                   The Beaches                  0        0   
2                   The Beaches                  0        0   
3                   The Beaches                  0        0   
4  The Danforth West, Riverdale                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  \
0                 0                   0      0         0           0   
1                 0                   0      0         0           0   
2                 0                   0      0         0           0   
3                 0                   0      0         0           0   
4                 0                   0      0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Cajun / Creole Restaurant  \
0             0              0        0     0                          0   
1             0              0        0     0                          0   
2             0              0        0     0                          0   
3             0              0        0     0                          0   
4             0              0        0     0                          0   

   Camera Store  Caribbean Restaurant

In [35]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                            Adelaide, King, Richmond           0.000000   
1                                         Berczy Park           0.000000   
2        Brockton, Exhibition Place, Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
5                         Cabbagetown, St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7           Chinatown, Grange Park, Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011628   
10                     Commerce Court, Victoria Hotel           0.000000   
11           Design Exchange, Toronto Dominion Centre           0.000000   
12                       Dovercourt Village, Dufferin           0.000000   
13             First Canadian Place, Underground city           0.000000   
14                     Harbord, University of Toronto           0.000000   
15  Harbourfront East, Toronto Islands, Union Station           0.000000   
16                          Harbourfront, Regent Park           0.000000   
17                      High Park, The Junction South           0.000000   
18                           Little Portugal, Trinity           0.000000   
19                             Parkdale, Roncesvalles           0.000000   
20                                           Rosedale           0.000000   
21                                 Runnymede, Swansea           0.000000   
22                           Ryerson, Garden District           0.000000   
23                                     St. James Town           0.000000   
24                    Stn A PO Boxes 25 The Esplanade           0.000000   
25                                    Studio District           0.000000   
26                                        The Beaches           0.000000   
27                     The Beaches West, India Bazaar           0.000000   
28                       The Danforth West, Riverdale           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0000              0.0000        0.0000           0.000   
4    0.0625              0.0625        0.0625           0.125   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000              0.0000        0.0000           0.000   
15   0.0000              0.0000        0.0000           0.000   
16   0.0000              0.0000        0.0000           0.000   
17   0.0000              0.0000        0.0000           0.000   
18   0.0000              0.0000        0.0000           0.000   
19   0.0000              0.0000        0.0000           0.000   
20   0.0000              0.0000        0.0000           0.000   
21   0.0000              0.0000        0.0000           0.000   
22   0.0000              0.0000        0.0000           0.000   
23   0.0000              0.0000        0.0000           0.000   
24   0.0000              0.0000        0.0000           0.000   
25   0.0000              0.0000        0.0000           0.000   
26  

In [36]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3  Thai Restaurant  0.04
4       Steakhouse  0.04


----Berczy Park----
          venue  freq
0   Coffee Shop  0.11
1  Cocktail Bar  0.05
2   Cheese Shop  0.04
3          Café  0.04
4      Beer Bar  0.04


----Brockton, Exhibition Place, Parkdale Village----
                   venue  freq
0            Coffee Shop  0.09
1         Breakfast Spot  0.09
2                   Café  0.09
3  Performing Arts Venue  0.09
4             Restaurant  0.04


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1                 Spa  0.06
2       Auto Workshop  0.06
3         Pizza Place  0.06
4       Burrito Place  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12

In [37]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [38]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Café                   Bar       Thai Restaurant   
1           Cocktail Bar              Beer Bar        Farmers Market   
2  Performing Arts Venue                  Café           Coffee Shop   
3            Pizza Place         Auto Workshop            Comic Shop   
4         Airport Lounge      Airport Terminal              Boutique   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                   Hotel            Restaurant   
1                Bakery              Steakhouse           Cheese Shop   
2                   Gym  Furniture / Home Store    Falafel Restaurant   
3      Recording Studio              Restaurant               Butcher   
4               Airport      Airport Food Court          Airport Gate   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint   American Restaurant       Asian Restaurant  
1    Seafood Restaurant                  Café     Belgian Restaurant  
2     Convenience Store         Burrito Place         Sandwich Place  
3         Burrito Place               Brewery             Skate Park  
4       Harbor / Marina                   Bar       Sculpture Garden

In [39]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 2, 0, 0, 0, 4, 0], dtype=int32)

In [40]:
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)


In [46]:
neighbourhoods_venues_sorted.head()

Cluster Labels                                      Neighbourhood  \
0               0                           Adelaide, King, Richmond   
1               0                                        Berczy Park   
2               0       Brockton, Exhibition Place, Parkdale Village   
3               0  Business Reply Mail Processing Centre 969 Eastern   
4               2  CN Tower, Bathurst Quay, Island airport, Harbo...   

  1st Most Common Venue  2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                   Café                   Bar   
1           Coffee Shop           Cocktail Bar              Beer Bar   
2        Breakfast Spot  Performing Arts Venue                  Café   
3    Light Rail Station            Pizza Place         Auto Workshop   
4       Airport Service         Airport Lounge      Airport Terminal   

  4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
0       Thai Restaurant            Steakhouse                   Hotel   
1        Farmers Market                Bakery              Steakhouse   
2           Coffee Shop                   Gym  Furniture / Home Store   
3            Comic Shop      Recording Studio              Restaurant   
4              Boutique               Airport      Airport Food Court   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Restaurant          Burger Joint   American Restaurant   
1           Cheese Shop    Seafood Restaurant                  Café   
2    Falafel Restaurant     Convenience Store         Burrito Place   
3               Butcher         Burrito Place               Brewery   
4          Airport Gate       Harbor / Marina                   Bar   

  10th Most Common Venue  
0       Asian Restaurant  
1     Belgian Restaurant  
2         Sandwich Place  
3             Skate Park  
4       Sculpture Garden

In [48]:
# add clustering labels

Toronto_merged = Toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = Toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')


In [51]:
toronto_merged.head()

Postal Code           Borough                   Neighbourhood   Latitude  \
0         M4E      East Toronto                     The Beaches  43.676357   
1         M4K      East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L      East Toronto  The Beaches West, India Bazaar  43.668999   
3         M4M      East Toronto                 Studio District  43.659526   
4         M4W  Downtown Toronto                        Rosedale  43.679563   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               3     Health Food Store                   Pub   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0        Sandwich Place         Burrito Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.377529               1                  Park            Playground   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                 Trail          Neighborhood          Cosmetics Shop   
1        Ice Cream Shop    Italian Restaurant  Furniture / Home Store   
2     Food & Drink Shop            Steakhouse       Fish & Chips Shop   
3    Italian Restaurant                Bakery     American Restaurant   
4                 Trail              Building          Discount Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0       Coworking Space        Farmers Market    Falafel Restaurant   
1           Pizza Place             Bookstore               Brewery   
2  Fast Food Restaurant               Brewery          Burger Joint   
3             Gastropub           Cheese Shop           Fish Market   
4        Farmers Market    Falafel Restaurant           Event Space   

  9th Most Common Venue     10th Most Common Venue  
0           Event Space       Ethiopian Restaurant  
1       Bubble Tea Shop       Caribbean Restaurant  
2                   Gym           Sushi Restaurant  
3             Bookstore  Latin American Restaurant  
4  Ethiopian Restaurant          Electronics Store

In [52]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Folium installed and imported!')

Solving environment: done

# All requested packages already installed.

Folium installed and imported!


In [53]:
address = 'Toronto'

geolocator = Nominatim(user_agent="To_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [54]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters